nucypher/characters/base.py

```diff
@@ -152,46 +159,52 @@ def __init__(self,
             self._crypto_power = CryptoPower(power_ups=self._default_crypto_powerups)
 
         #
-        # Self-Character
+        # Self
         #
 
         if is_me:
-            #
+
             # Signing Power
-            #
             self.signer = signer
             try:
                 signing_power = self._crypto_power.power_ups(SigningPower)  # type: SigningPower
                 self._stamp = signing_power.get_signature_stamp()  # type: SignatureStamp
             except NoSigningPower:
                 self._stamp = NO_SIGNING_POWER
 
-            #
             # Blockchain
-            #
-            self.provider_uri = provider_uri
             if not self.federated_only:
                 self.registry = registry or InMemoryContractRegistry.from_latest_publication(network=domain)  # See #1580
             else:
                 self.registry = NO_BLOCKCHAIN_CONNECTION.bool_value(False)
 
+            if not federated_only and not provider_uri:
+                raise ValueError('Provider URI is required to init a decentralized character.')
+            self.provider_uri = provider_uri
+
             # REST
             self.network_middleware = network_middleware or RestMiddleware(registry=self.registry)
 
-            #
             # Learner
-            #
+            self.suspicious_activities_witnessed = defaultdict(list)  # TODO: Combine with buckets / node labeling
             Learner.__init__(self,
                              domain=domain,
                              network_middleware=self.network_middleware,
                              node_class=known_node_class,
                              *args, **kwargs)
 
+            if self.federated_only:
+                try:
+                    checksum_address = self.derive_federated_address()
+                except NoSigningPower:
+                    checksum_address = NO_SIGNING_POWER.bool_value(False)
```

Comment:
Seems like we are overwriting the value of `checksum_address` from `__init__` here as well. What if `checksum_address` was already provided in the `__init__`? Since this is federated, do we need a check to make sure the checksum_address if provided in `__init__` matches the derived address from the public verifying key?

Reply:
Yes - Ultimately the character's stamp is the correct material for federated address derivation.  If something else is passed, it's not valid.

Reply:
✔️